csi/src/error.rs

```diff
@@ -0,0 +1,290 @@
+        /// Node ID
+        node_id: String,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Argument is invalid
+    #[error("Argument is invalid, context is {:?}", .context)]
+    ArgumentInvalid {
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Argument is out of range
+    #[error("Argument is out of range, context is {:?}", .context)]
+    ArgumentOutOfRange {
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by std::path::StripPrefixError
+    #[error("StripPrefixErr, the error is {:?}, context is {:?}", .source, .context)]
+    StripPrefixErr {
+        /// Error source
+        source: std::path::StripPrefixError,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by etcd_client::Error,
+    #[error("EtcdClientErr, the error is {:?}, context is {:?}", .source, .context)]
+    EtcdClientErr {
+        /// Error source
+        source: etcd_client::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by bincode::Error
+    #[error("BincodeErr, the error is {:?}, context is {:?}", .source, .context)]
+    BincodeErr {
+        /// Error source
+        source: bincode::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by nix::Error
+    #[error("NixErr, the error is {:?}, context is {:?}", .source, .context)]
+    NixErr {
+        /// Error source
+        source: nix::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Failed to mount
+    #[error("MountErr, fail to mount {:?} to {:?}, context is {:?}", .from, .target, .context)]
+    MountErr {
+        /// Source to mount
+        from: PathBuf,
+        /// Mount point
+        target: PathBuf,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Failed to umount
+    #[error("UmountErr, fail to umount {:?}, context is {:?}", .target, .context)]
+    UmountErr {
+        /// Mount point to umount
+        target: PathBuf,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by std::time::SystemTimeError
+    #[error("SystemTimeErr, the error is {:?}, context is {:?}", .source, .context)]
+    SystemTimeErr {
+        /// Error source
+        source: std::time::SystemTimeError,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// Error caused by grpcio::Error
+    #[error("GrpcioErr, the error is {:?}, context is {:?}", .source, .context)]
+    GrpcioErr {
+        /// Error source
+        source: grpcio::Error,
+        /// Context of the error
+        context: Vec<String>,
+    },
+
+    /// API is not implemented
+    #[error("Not implemented, context is {:?}", .context)]
+    Unimplemented {
+        /// Context of the error
+        context: Vec<String>,
```

Comment:
As context is in all enum variant, can we implement it by some macro?

Reply:
Using macro here will break the readability. It's not necessary.

Reply:
Yes. Agree.